### PR TITLE
Update RapidsUDF interface to support UDFs with no input parameters

### LINKED ISSUE #7373: [BUG] RapidsUDF does not support UDFs with no inputs
The RapidsUDF interface, which is used to implement RAPIDS accelerated UDFs, does not support UDFs that do not take any inputs but still generate a value (i.e.: a generator-like UDF).  The evaluateColumnar method will be called with an empty array of ColumnVector inputs, and the columnar method has no way to determine how many rows to emit in the resulting ColumnVector output.

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #7373 by deprecating the old interface and providing a new interface that passes the number of rows as the first argument.  The new method has a default implementation that calls the original, deprecated method for backwards compatibility, and I verified that UDFs compiled against the original interface before this change continue to work against the new interface at runtime.

### CODE REVIEW

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
Filed NVIDIA/spark-rapids-examples#258 to track updating the UDF examples to stop using the deprecated method and use the new method.

File: sql-plugin/src/main/java/com/nvidia/spark/RapidsUDF.java [hunk 1]
```diff
@@ -35,6 +37,34 @@ public interface RapidsUDF {
...
+
+  /**
+   * Evaluate a user-defined function with RAPIDS cuDF columnar inputs
+   * producing a cuDF column as output. The method must return a column of
+   * the appropriate type that corresponds to the type returned by the CPU
+   * implementation of the UDF (e.g.: INT32 for int, FLOAT64 for double,
+   * STRING for String, etc) or a runtime exception will occur when the
+   * results are marshalled into the expected Spark result type for the UDF.
+   * The number of rows of output must match the number of rows specified,
+   * and all input columns will have that same number of rows.
```
Comment (gerashegalov) [inline sql-plugin/src/main/java/com/nvidia/spark/RapidsUDF.java:56]:
```suggestion
   * and all input columns must have that same number of rows.
```

**Review by gerashegalov — APPROVED** [review]
LGTM

**Comment by jlowe** [comment]
build

**Review by abellina — APPROVED** [review]

**Review by gerashegalov — APPROVED** [review]

File: tests/src/test/scala/com/nvidia/spark/rapids/tests/udf/scala/URLDecode.scala [hunk 2]
```diff
@@ -40,11 +40,12 @@ class URLDecode extends Function[String, String] with RapidsUDF with Serializabl
   }
 
   /** Columnar implementation that runs on the GPU */
-  override def evaluateColumnar(args: ColumnVector*): ColumnVector = {
+  override def evaluateColumnar(numRows: Int, args: ColumnVector*): ColumnVector = {
     // The CPU implementation takes a single string argument, so similarly
     // there should only be one column argument of type STRING.
     require(args.length == 1, s"Unexpected argument count: ${args.length}")
     val input = args.head
+    require(numRows == input.getRowCount, s"Expected $numRows rows, received ${input.getRowCount}")
```
Comment (GaryShen2008) [inline tests/src/test/scala/com/nvidia/spark/rapids/tests/udf/scala/URLDecode.scala:48]: nit: Why need to check the numRows's correctness, but no where to use it in URLDecode case?
It should be ok to skip this line, if URLDecode won't use numRows.
I think in most case of using RapidsUDF, we just need to add the numRows parameter and the logic should be as same as before. Only the case what you mentioned in the issue, which numRows has number but no data in the input ColumnVector, we should use numRows to do something like creating a scalar.
Am I right？

Comment (jlowe) [reply tests/src/test/scala/com/nvidia/spark/rapids/tests/udf/scala/URLDecode.scala:48]: Given this UDF is part of a unit test for RAPIDS accelerated UDFs, the assert is warranted.  If we ever pass an incorrect value here, that would be quite bad.  The point of checking it here is to validate the framework is doing the right thing from the user's perspective (i.e.: the one implementing the RAPIDS accelerated UDF code).

**Review by revans2 — APPROVED** [review]

**Pull request merged**